### PR TITLE
refactor: use `derive` interface of `clap`

### PR DESCRIPTION
**Pull request by ErichDonGubler**
Hey there! 👋🏻 I figured this would be a nice way to make editing the CLI more approachable, since this only takes on 1 new dependency ([`heck` 0.5](https://docs.rs/heck/0.5.0/heck/)). LMK what you think!

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,77 +1,56 @@
...
 
-use clap::ArgAction;
+use clap::{CommandFactory, Parser as _};
 use clap_complete::{generate, Shell};
 use clap_complete_nushell::Nushell;
 use slog::Drain;
 use std::io;
 
+#[derive(Debug, clap::Parser)]
+#[clap(about = "Automatically absorb staged changes into your current branch")]
```
Comment (tummychow) [inline src/main.rs:11]: can this be specified as a doc comment?

File: src/main.rs [hunk 2]
```diff
@@ -1,77 +1,56 @@
...
+struct Cli {
+    /// Use this commit as the base of the absorb stack
+    #[clap(long, short, value_name = "base")]
```
Comment (tummychow) [inline src/main.rs:14]: what does `--help` output look like if we omit these `value_name`s? they're mildly redundant

File: src/main.rs [hunk 3]
```diff
@@ -1,77 +1,56 @@
-#[macro_use]
-extern crate clap;
-
 #[macro_use]
 extern crate slog;
 
-use clap::ArgAction;
+use clap::{CommandFactory, Parser as _};
```
Comment (tummychow) [inline src/main.rs:4]: is commandfactory necessary here? most of the flag-only examples don't seem to require it

**Review by tummychow — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 1 above)
Comment (ErichDonGubler) [reply src/main.rs]: Yes! I've changed it to be a doc. comment (which seems your implicit request) with b90d9b84db326bbf63b4051a6d45900de44fa0f9, which can be [autosquashed](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---autosquash).

File: src/main.rs (hunk 2 above)
Comment (ErichDonGubler) [reply src/main.rs:14]: I only put them in there to ensure complete parity with the old `--help`. If having these be `SCREAMING_SNAKE_CASE`'d is acceptable, then we can remove them; that's the only difference.

File: src/main.rs (hunk 3 above)
Comment (ErichDonGubler) [reply src/main.rs:4]: It's so we can call `Cli::command` to feed a `Command` structure to `clap_completion::generate` below. If one removes the import, they get the following compiler error:

```
error[E0599]: no function or associated item named `command` found for struct `Cli` in the current scope
  --> src/main.rs:53:35
   |
12 | struct Cli {
   | ---------- function or associated item `command` not found for this struct
...
53 |         let mut args_clone = Cli::command();
   |                                   ^^^^^^^ function or associated item not found in `Cli`
   |
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
   |
1  + use clap::CommandFactory;
   |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `git-absorb` (bin "git-absorb") due to 1 previous error
```

File: src/main.rs (hunk 2 above)
Comment (ErichDonGubler) [reply src/main.rs]: I've preemptively pushed 57fa171 to remove these `value_name` attributes. Feel free to rebase interactively and `drop` this commit if you decide you'd rather not change `help` test. 🙂

File: src/main.rs (hunk 3 above)
Comment (ErichDonGubler) [reply src/main.rs:4]: I consider this resolved, but I'll leave this open for visibility.

**Comment by ErichDonGubler** [comment]
I've `autosquash`'d the feedback commits I've linked to in earlier conversations, and force-pushed the "clean" resulting history.

**Comment by kiprasmel** [comment]
@ErichDonGubler i want to implement `--foo` and `--no-foo`. previously before this PR, i could simply get the last index of both args (if any index), and whichever index was bigger, that value of either `--foo` or `--no-foo` would prevail.

how would I implement this now?

**Comment by ErichDonGubler** [comment]
@kiprasmel: It's still possible, though it might look a bit odd! I didn't realize I had affected you here. 😅

If you're determined to use the `--{,no-}option` scheme, you could something like `clap`'s tests for specification on an `overrides_with` option, like in [`clap`'s tests upstream](https://github.com/clap-rs/clap/blob/9d14f394ba22f65f8957310a03ae5fd613f89d76/tests/derive/flags.rs#L284-L330) or [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0715e1a76358900b65f7bc4f527ba34f) I made to demonstrate; in short:

```rust
#[derive(Debug, Eq, clap::Parser, PartialEq)]
struct Cli {
    #[clap(long)]
    foo: bool,
    #[arg(long, action = clap::ArgAction::SetTrue, overrides_with = "foo")]
    no_foo: (),
}
```

In case it's interesting, I've often found that I prefer using a doubly-optional `ValueEnum` with a variant that disables the flag, for forwards compatibility and expressiveness. I don't know if the suggestion is relevant to you (you're probably dealing with an actually `bool`ean solution space), but you might find this an overall more clear alternative, depending on what you're implementing ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=61a545c9c9e3056c6c7ee8d5125ad020)):

```rust
#[derive(Debug, Eq, clap::Parser, PartialEq)]
struct Cli {
    #[clap(long)]
    // One can remove the outer `Option` to force specification of the value on
    // the CLI.
    foo: Option<Option<Foo>>,
}

#[derive(Clone, Debug, Eq, PartialEq, clap::ValueEnum)]
enum Foo {
    Yes,
    No,
    Whatever,
}
```

**Comment by kiprasmel** [comment]
@ErichDonGubler yes I thought about having the foo and no_foo as cli args, and resolve them into final value when passing in the config to run absorb.

but again, i want a new --foo or --no-foo to overwrite the previous one, thus need to access the index.

is there a way to simply access the argv strings and simply do it myself?

**Comment by tummychow** [comment]
for the record, having graceful precedence of `--foo`/`--not-foo` is low importance to me. definitely not important enough to write custom cli parsing code beyond what clap already does - in fact, i would even merge a change that only added `--foo` without implementing `--no-foo` at all. it's unfortunate that clap doesn't already support this (https://github.com/clap-rs/clap/issues/815) but i care more about having as little cli parsing code as possible

**Comment by ErichDonGubler** [comment]
@kiprasmel:

> @ErichDonGubler yes I thought about having the foo and no_foo as cli args, and resolve them into final value when passing in the config to run absorb.

If you examine the cases in both the tests and the playground link, I believe even the question of precedence is handled for you; AFAICT, the last flag specified "wins". Am I not understanding your intent?

**Pull request merged**